File: binary_search_tree/BinarySearchTree.java
import java.util.NoSuchElementException;

class Node {	// Node for Binary Search Tree
    public int data;
    public Node left;
    public Node right;

    public Node(int data) {	// Initializes node with given data and no chlid
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BST {
    private Node root;	// Root of Binary Search Tree

    public BST() {	// Initializes empty BST
        root = null;
    }

    public BST(int data) {	// Initializes root of BST
        root = new Node(data);
    }

    public void insert(int data) {
        if (root == null) {
            root = new Node(data);	// Initialize root with the given data
            return;
        }
        Node newNode = new Node(data);	// Create Node for new entry
        Node iterator = root;	// Temp Node for iterating from root to leaf
        Node parent = null;	// To store the future parent of new node
        while (iterator != null) {
            parent = iterator;
            if (data <= iterator.data) {
                iterator = iterator.left;
            }
            else {
                iterator = iterator.right;
            }
        }
        if (data <= parent.data) {
            parent.left = newNode;
        }
        else {
            parent.right = newNode;
        }
    }

    public Node search(int data) {	// Returns the Node if element is found else will throw an Exception
        Node iterator = root;
        while (iterator != null) {
            if (iterator.data == data)
                return iterator;
            else if (data <= iterator.data)
                iterator = iterator.left;
            else
                iterator = iterator.right;
        }
        throw new NoSuchElementException("Element is not found in BST");
    }

    public boolean delete(int data) {	// Finds the parent of the node to be deleted.
        if (root == null) {
            throw new NoSuchElementException("Cannot perform delete operation, BST is empty");
        }
        Node iterator = root;
        Node parent = null;
        while (iterator != null) {
            if (data == iterator.data) {
                return deleteNode(data, parent);
            } else  {
                parent = iterator;
                if (data <= iterator.data)
                    iterator = iterator.left;
                else
                    iterator = iterator.right;
            }
        }
        throw new NoSuchElementException("Delete Unsuccessful! Element was not found in BST");
    }

    private boolean deleteNode(int data, Node parent) {
        Node child = null;
        boolean position = false;	// Indicates position of child wrt to parent, true is left child, false is right child
        if (data <= parent.data) {
            child = parent.left;
            position = true;
        }
        else
            child = parent.right;

        if (child.left == child.right) {	// Condition for leaf node
            child = null;
            if (position)
                parent.left = null;
            else
                parent.right = null;
            return true;
        } else if (child.right == null) {	// Condition for non-leaf with no right sub-tree
            if (position)
                parent.left = child.left;
            else
                parent.right = child.left;
            child.left = null;
            child = null;
            return true;
        } else if (child.left == null) {	// Condition for non-leaf with no left sub-tree
            if (position)
                parent.left = child.right;
            else
                parent.right = child.right;
            child.right = null;
            child = null;
            return true;
        }
        else {	// Conditon when Node has both subtree avaliable
            Node iterator = child.right;
            Node parentOfIterator = null;
            while(iterator.left != null) {	// Finding the leftmost child of right sub-tree
                parentOfIterator = iterator;
                iterator = iterator.left;
            }
            child.data = iterator.data;
            parentOfIterator.left = null;
            iterator = null;
            return true;
        }
    }

    public void printInOrder() {	// Function to call inorder printing using root
        if (root == null)
            throw new NoSuchElementException("Cannot print! BST is empty");
        print(root);
        System.out.println("");
    }

    private void print(Node iterator) {
        if (iterator != null) {
            print(iterator.left);
            System.out.print(iterator.data + " ");
            print(iterator.right);
        }
    }
}

public class BinarySearchTree {
    public static void main(String[] args) {
        // Created an empty tree
        BST tree = new BST();
        // Adding a few test entries
        tree.insert(10);
        tree.insert(9);
        tree.insert(3);
        tree.insert(12);
        tree.insert(14);
        tree.insert(7);
        tree.insert(6);
        tree.insert(11);
        tree.insert(1);
        tree.insert(2);
        // Test printing
        tree.printInOrder();
        // Deleting a valid node
        tree.delete(9);
        // Print again
        tree.printInOrder();
        // Searching an invalid node, same can be tested for delete as both use same logic
        // but with a slight different approach to find the node
        try {
            tree.search(4);
            System.out.println("Node was found successfully.");
        } catch (NoSuchElementException e) {
            System.out.println("Invalid Search");
        }
        try {
            tree.delete(9);
        } catch (NoSuchElementException  e) {
            System.out.println("Cannot delete, Node not present.");
        }
    }
}

File: binary_search_tree/binary_search_tree.py
import sys


class Node(object):
    """
    All data used in BST stored as object of Node class
    data : It is the value store in object
    left : it is left child of the object
    right : It is right child of the object
    """

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class bst(object):

    def __init__(self):
        # here root is root of our BST tree which is initially None
        self.root = None

    def insert_node(self, data):
        """
        Add Node in binary tree
        Time Complexity: O(Logn)average case and O(n) in worst case
        :param data:Node with value data will add in BST
        """
        iterator = self.root
        new_node = Node(data)
        if(self.root is None):
            self.root = new_node
        else:
            while iterator is not None:
                if iterator.data >= data:
                    if iterator.left is None:
                        iterator.left = new_node
                        break
                    else:
                        iterator = iterator.left
                else:
                    if iterator.right is None:
                        iterator.right = new_node
                        break
                    else:
                        iterator = iterator.right
        return

    def delete_node(self, data):
        """
        Delete element from binary tree
        Time Complexity: O(Logn)average case and O(n) in worst case
        :param data: Node with value of data will delete from bst
        Throws execption if node with value of data not exist
        """
        try:
            parent_node = self.search_node(data)
            if parent_node is None:
                node = self.root
                if node.right is None and node.left is None:
                    # if node has no child
                    node = None
                elif node.right is None:
                    # if node only has left child
                    node = node.left
                elif node.left is None:
                    # if node only has right child
                    node = node.right
                else:
                    # if node have both children
                    if node.right.left is None:
                        node.right.left = node.left
                        node = node.right
                        self.root = node
                    else:
                        to_replace = node.right
                        parent_to_replace = node
                        while to_replace.left is not None:
                            parent_to_replace = to_replace
                            to_replace = to_replace.left
                        node.data = to_replace.data
                        parent_to_replace.left = to_replace.right
                        self.root = node
            else:
                if (parent_node.left is not None and
                        parent_node.left.data is data):
                    node = parent_node.left
                    if node.right is None and node.left is None:
                        # if node  has no child
                        parent_node.left = None
                    elif node.right is None:
                        # if node only has left child
                        parent_node.left = node.left
                    elif node.left is None:
                        # if node only has right child
                        parent_node.left = node.right
                    else:
                        # if node have both children
                        if node.right.left is None:
                            node.right.left = node.left
                            parent_node.left = node.right
                        else:
                            to_replace = node.right
                            parent_to_replace = node
                            while to_replace.left is not None:
                                parent_to_replace = to_replace
                                to_replace = to_replace.left
                            parent_node.left.data = to_replace.data
                            parent_to_replace.left = to_replace.right
                    return
                else:
                    node = parent_node.right
                    if node.right is None and node.left is None:
                        # if node has no child
                        parent_node.right = None
                    elif node.right is None:
                        # if node only has left child
                        parent_node.right = node.left
                    elif node.left is None:
                        # if node only has right child
                        parent_node.right = node.right
                    else:
                        # if node have both chidren
                        if node.right.left is None:
                            node.right.left = node.left
                            parent_node.right = node.right
                        else:
                            to_replace = node.right
                            parent_to_replace = node
                            while to_replace.left is not None:
                                parent_to_replace = to_replace
                                to_replace = to_replace.left
                            parent_node.right.data = to_replace.data
                            parent_to_replace.left = to_replace.right
                    return
        except LookupError as e:
            print('sorry,%s' % (e))

    def search_node(self, data):
        """
        search element in binary tree
        Time Complexity: O(Logn)average case and O(n) in worst case
        :param data: search node of value data in bst
        :return : parent of the node whose value is data or raise exception
        """
        parent_node = None
        if self.root is None:
            raise LookupError('Cannot delete anything, BST is empty')
        elif self.root.data == data:
            return parent_node
        else:
            iterator = self.root
            while iterator is not None:
                if iterator.data == data:
                    return parent_node
                elif iterator.data > data:
                    parent_node = iterator
                    iterator = iterator.left
                else:
                    parent_node = iterator
                    iterator = iterator.right
            raise LookupError('Node does not exist')

    def print_preorder(self, node):
        """
        Print all element of binary tree in preorder using recurtion
        """
        if node is None:
            return
        elif node.left is None and node.right is None:
            sys.stdout.write(str(node.data) + ' ')
        else:
            self.print_preorder(node.left)
            sys.stdout.write(str(node.data) + ' ')
            self.print_preorder(node.right)
            if self.root == node:
                sys.stdout.write('\n')
            return


def main():
    inp = bst()
    inp.delete_node(2)  # try to delete node in empty try
    inp.insert_node(4)
    inp.insert_node(8)
    inp.insert_node(6)
    inp.insert_node(5)
    inp.insert_node(44)
    inp.insert_node(69)
    inp.print_preorder(inp.root)
    inp.delete_node(69)
    inp.delete_node(5)
    inp.delete_node(4)  # delete root
    inp.print_preorder(inp.root)
    print(inp.root.data)  # check value of root
    inp.delete_node(2)  # just to throw exception


if __name__ == '__main__':
    main()
